Fix dropped NodePort traffic to hostNetwork backends with Geneve+DSR

tommasopozzetti opened the pull request:
This PR fixes DSR dispatch with Geneve to hostNetworked pods by ensuring tunnel_endpoint is always populated in ipcache, even for directly reachable endpoints. Unneeded encapsulation is avoided by passing the skiptunnel flag for such endpoints.

Fixes: #36901

Artyop reviewed (approved):
lgtm

julianwiedmann commented on bpf/lib/nodeport.h:
```diff
@@ -1617,12 +1617,16 @@ static __always_inline int encap_geneve_dsr_opt4(struct __ctx_buff *ctx, int l3_
+		tunnel_endpoint = ip4->daddr;
+	else
+		tunnel_endpoint = info->tunnel_endpoint;
+
```
Let's preserve old behavior here

```suggestion
	if (info->tunnel_endpoint)
		tunnel_endpoint = info->tunnel_endpoint;
	else if (identity_is_remote_node(dst_sec_identity))
	        tunnel_endpoint = ip4->daddr;
	else
		return DROP_NO_TUNNEL_ENDPOINT;

```

julianwiedmann commented on cilium-cli/connectivity/builder/north_south_loadbalancing.go:
```diff
@@ -16,5 +16,8 @@ func (t northSouthLoadbalancing) build(ct *check.ConnectivityTest, _ map[string]
+		WithScenarios(
+			tests.OutsideToNodePort(),
+			tests.OutsideToNodePortForHostNetNs(),
+		)
```
For compatibility we'll need some variant of `WithCiliumVersion(">=1.18.0")` here, until the backports to older Cilium versions have landed.

julianwiedmann commented on bpf/lib/nodeport.h:
```diff
@@ -1617,12 +1617,16 @@ static __always_inline int encap_geneve_dsr_opt4(struct __ctx_buff *ctx, int l3_
 	dst_sec_identity = info->sec_identity;
 
+	if (identity_is_remote_node(dst_sec_identity))
+		tunnel_endpoint = ip4->daddr;
```
One unresolved question would be how this works for IPv6 (in `encap_geneve_dsr_opt6()`). We currently don't support overlay network over IPv6, so there we can't easily fall back to the backend IP :grimacing:.

The big hammer would be to touch https://github.com/cilium/cilium/blob/3d75eba2e41d6c0fd839feb4b4eb99904250edb3/pkg/node/manager/manager.go#L567. In native-routing that would nicely address this problem. But in overlay mode it would cause *all* `to-remote-node` traffic to flow via the overlay network. Which is something we've avoided so far ...

julianwiedmann reviewed (commented):
Some comments inline, before I forgot about the CLI compatibility aspect again ...

dylandreimerink reviewed (changes requested):
> One unresolved question would be how this works for IPv6 (in encap_geneve_dsr_opt6()). We currently don't support overlay network over IPv6, so there we can't easily fall back to the backend IP 😬.

I think its important to address this and not make the IPv4 and IPv6 paths even more asymmetrical then it already is.

I suggest that we change the logic in the agent to always populate the `tunnel_endpoint` field. In addition also set the `flag_skip_tunnel` [flag](https://github.com/cilium/cilium/pull/29796). This flag should prevent regular tunneling to the host endpoints (unwanted behavior @julianwiedmann pointed out) but still gives us access for the purposes of DSR. Since current Cilium versions already respect the flag, except for this DSR logic, this should also be safe during upgrades and downgrades as well. We would have to make sure to not set this flag when the host firewall feature is enabled.

jschwinger233 reviewed (approved):
encryption parts LGTM.

(my personal feeling is, maybe we can split it into two commits, one for "always set tunnel_endpoint in ipcache", the other for "propagate endpointFlags")

derailed reviewed (approved):
@tommasopozzetti Nice work!

thorn3r reviewed (approved):
agent changes LGTM :+1:

sayboras reviewed (approved):
Thanks, the changes look good for envoy related file (e.g. pkg/envoy/resources.go)

tommasopozzetti committed:
Add plumbing to set and propagate endpointFlags from #29796

This commit makes endpoint flags available for plumbing in the new
ipcache API, allowing to set them from higher in the stack

Signed-off-by: Tommaso Pozzetti <tommypozzetti@hotmail.it>

tommasopozzetti committed:
Fix NodePort traffic to host netns backends with Geneve+DSR

Always set tunnel_endpoint in ipcache, even for directly reachable
endpoints. This ensures DSR info is available for DSR with geneve
dispatch.
Use endpointFlags to set the skiptunnel flag for directly reachable
endpoints such that, even if overlay is enabled, we do not do
an unneeded encap.

Fixes: #36901

Signed-off-by: Tommaso Pozzetti <tommypozzetti@hotmail.it>

doniacld reviewed (approved):
LGTM but asking peers from [cilium/ipcache](https://github.com/orgs/cilium/teams/ipcache) to have a second look regarding the addition of a new field to `IPCache`. cc @squeed
